ethers-signers/src/wallet/mod.rs

```diff
@@ -83,6 +87,23 @@ impl<D: Sync + Send + DigestSigner<Sha256Proxy, RecoverableSignature>> Signer fo
         Ok(self.sign_transaction_sync(tx))
     }
 
+    async fn sign_typed_data<T: Eip712 + Send + Sync>(
+        &self,
+        payload: &T,
+    ) -> Result<Option<Signature>, Self::Error> {
+        if let Ok(encoded) = payload.encode_eip712() {
```

Comment:
Why can you not do `payload.encode_eip712()?` etc for the other methods?

Reply:
Ideally, we could call `payload.encode_eip712()?`, but I am still trying to figure out how to map the error to the `std::convert::Infallible` type. Would changing the trait Error type for the wallet signer cause any downstream issues? It seems like there could be cases where an error may happen, which doesn't seem infallible to me.

Reply:
I've replaced `std::convert::Infallible` with `WalletError`: 

`type Error = WalletError;`

Added an enum for the `<T as Eip712>::Error` type for mapping the error to WalletError:

```rust
async fn sign_typed_data<T: Eip712 + Send + Sync>(
        &self,
        payload: T,
    ) -> Result<Signature, Self::Error> {
        let encoded = payload
            .encode_eip712()
            .map_err(|e| Self::Error::Eip712Error(e.to_string()))?;

        Ok(self.sign_hash(H256::from(encoded), false))
    }
```

Decided against implementing a trait bound on `WalletError` to satisfy `<T as Eip712>::Error` and decided to use a message string:

```rust
pub enum WalletError {

    ...

    /// Error type from Eip712Error message
    #[error("error encoding eip712 struct: {0:?}")]
    Eip712Error(String),
}
```

Also open to other suggestions for better error handling.